File: sdk/java/src/main/java/com/anyvali/schemas/TupleSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Schema for fixed-length tuples with positional element schemas.
 */
public class TupleSchema extends Schema {
    private final List<Schema> items;

    public TupleSchema(List<Schema> items) {
        this.items = new ArrayList<>(items);
    }

    private TupleSchema(TupleSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.items = new ArrayList<>(other.items);
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        if (!(input instanceof List<?> list)) {
            String received = anyvaliTypeName(input);
            ctx.addIssue(IssueCodes.INVALID_TYPE,
                    "Expected tuple, received " + received, "tuple", received);
            return null;
        }

        if (list.size() != items.size()) {
            String code = list.size() < items.size() ? IssueCodes.TOO_SMALL : IssueCodes.TOO_LARGE;
            ctx.addIssue(code,
                    "Expected exactly " + items.size() + " items, received " + list.size(),
                    items.size(), list.size());
            return null;
        }

        var result = new ArrayList<>();
        for (int i = 0; i < items.size(); i++) {
            var childCtx = ctx.child(i);
            Object parsed = items.get(i).runPipeline(list.get(i), childCtx);
            result.add(parsed);
        }
        return result;
    }

    @Override
    protected Map<String, Object> toNode() {
        var elements = new ArrayList<Map<String, Object>>();
        for (Schema item : items) {
            elements.add(item.toNode());
        }
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", "tuple");
        node.put("elements", elements);
        return addCommonNodeFields(node);
    }

    @Override
    protected Schema copy() {
        return new TupleSchema(this);
    }

    public List<Schema> getItems() {
        return items;
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/TupleSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

data class TupleSchema(
    val elements: List<Schema>
) : Schema() {
    override val kind: String = "tuple"

    override fun safeParseWithContext(input: Any?, ctx: ValidationContext): ParseResult {
        if (input !is List<*>) {
            return ParseResult.Failure(
                listOf(
                    ValidationIssue(
                        code = IssueCodes.INVALID_TYPE,
                        path = ctx.path,
                        expected = "tuple",
                        received = getJsonTypeName(input)
                    )
                )
            )
        }

        val issues = mutableListOf<ValidationIssue>()

        if (input.size < elements.size) {
            issues.add(
                ValidationIssue(
                    code = IssueCodes.TOO_SMALL,
                    path = ctx.path,
                    expected = elements.size.toString(),
                    received = input.size.toString()
                )
            )
            return ParseResult.Failure(issues)
        }
        if (input.size > elements.size) {
            issues.add(
                ValidationIssue(
                    code = IssueCodes.TOO_LARGE,
                    path = ctx.path,
                    expected = elements.size.toString(),
                    received = input.size.toString()
                )
            )
            return ParseResult.Failure(issues)
        }

        val result = mutableListOf<Any?>()
        for ((index, schema) in elements.withIndex()) {
            val childCtx = ctx.child(index)
            val childResult = schema.safeParseWithContext(input[index], childCtx)
            when (childResult) {
                is ParseResult.Success -> result.add(childResult.value)
                is ParseResult.Failure -> issues.addAll(childResult.issues)
            }
        }

        return if (issues.isEmpty()) ParseResult.Success(result) else ParseResult.Failure(issues)
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        val result = safeParseWithContext(value, ctx)
        return if (result is ParseResult.Failure) result.issues else emptyList()
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("tuple"))
        put("elements", JsonArray(elements.map { it.exportNode() }))
    }
}
